JBPM-8168: switch between case and process

xiezhang7 opened the pull request:
Assembly:
https://github.com/kiegroup/appformer/pull/709
https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/958
https://github.com/kiegroup/kie-wb-common/pull/2650
https://github.com/kiegroup/kie-wb-distributions/pull/927

xiezhang7 commented:
Assembly:
https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/958
https://github.com/kiegroup/kie-wb-common/pull/2650
https://github.com/kiegroup/kie-wb-distributions/pull/927

xiezhang7 commented:
Hi @romartin @LuboTerifaj , this is the _**new**_ pull request for [JBPM-8168](https://issues.jboss.org/browse/JBPM-8168). Thanks.

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Jenkins execute full downstream build

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/ManagedClientSessionCommands.java:
```diff
@@ -101,6 +101,10 @@ public void destroy() {
         sessionCommands.destroyAll();
     }
 
+    public void destroyCommands() {
```
mmm I see the (new) `destroyCommands` and (existing) `clearCommands` methods a bit redundant and confusing... 
Does it makes sense to replace the line `109` by line `105`, and so you can just call `clearCommands` (and so remove the `destroyCommands` too) ?

xiezhang7 replied:
@romartin I need `destroyCommands` method to destroy the instances of the commands. But `clearCommands` doesn't destroy the instances. I have changed [destroyCommands](https://github.com/kiegroup/kie-wb-common/blob/780c425886a399148819d023e5cb955df29dc707/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/ManagedClientSessionCommands.java#L104) to call 'clearCommands' and [destroy](https://github.com/kiegroup/kie-wb-common/blob/780c425886a399148819d023e5cb955df29dc707/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/ManagedClientSessionCommands.java#L100) method to call `destroyCommands`, so that there won't be any redundancy.

romartin replied:
yeahh well I was saying this because I think actually (on master) there exist a bug here in  
 the `clearCommands` method, as it's not calling the `sessionCommands#destroy(instance)`, as you did. So I was proposing just merging your code (fixed) into acutal `clearCommands` and reuse it.
Does that makes sense?

xiezhang7 replied:
Merged the code into method [clearCommands](https://github.com/kiegroup/kie-wb-common/blob/275abfeab02eae654f009ba2ec641fb39d3092e5/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/ManagedClientSessionCommands.java#L105). Thanks.

romartin replied:
thanks! :+1:

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-project-client/src/main/java/org/kie/workbench/common/stunner/cm/project/client/editor/CaseManagementDiagramEditor.java:
```diff
@@ -170,4 +258,88 @@ public boolean onMayClose() {
+
+    protected void onSwitch(final Diagram diagram, final String defSetId, final String shapeDefId,
+                            final Optional<SessionEditorPresenter<EditorSession>> sessionEditorPresenter) {
+        this.processView.showLoading();
```
It looks like this is not working completely for some reason, or probably it's still missing to manage the loading view properly on the CM editor side.  
Lemme explain - the main issue I see when testing it is that once switching between views, the users can still see the "old" version of a process for a while, and then suddently the canvas is being updated with latest state. 
So probably this processView is working fine, not completely sure, and then we still need to make it appear later, or is it a bug here?

romartin replied:
I reported this comment as an issue, on the review summary above

xiezhang7 replied:
@romartin With the new switch buttons, the view won't show the old version. It will only show after the backend finishes the switch.  
And it is supposed to show the loading screen as soon as the buttons are clicked: [this.showLoadingViews()](https://github.com/kiegroup/kie-wb-common/blob/780c425886a399148819d023e5cb955df29dc707/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-project-client/src/main/java/org/kie/workbench/common/stunner/cm/project/client/editor/CaseManagementDiagramEditor.java#L266).

![Peek 2019-05-23 22-44](https://user-images.githubusercontent.com/42279951/58299226-7f979c00-7dac-11e9-8b25-40dc2077a026.gif)

romartin replied:
ok, if the UI is working propertly, nothing to complain about :)

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-project-backend/src/main/java/com/kie/workbench/common/stunner/cm/service/CaseManagementSwitchViewServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+                    ((AbstractMetadata) metadata).setDefinitionSetId(mappedDefSetId);
+
+                    // Unmarshall the diagram
+                    try (final InputStream inputStream = new ByteArrayInputStream(rawData.getBytes())) {
```
I would just delegate to the framework (appformer) the reading of the file stream, otherwise we should handle some edge cases (long files, etc), I mean is the `ByteArrayInputStream` the best to consume the file? let's delegate it to our fundation libraries...
So I would recomment reading the bytes as, for example, [AbstractVFSDiagramService](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-backend-common/src/main/java/org/kie/workbench/common/stunner/core/backend/service/AbstractVFSDiagramService.java#L326) does, probably even you can change the [loadPath](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-backend-common/src/main/java/org/kie/workbench/common/stunner/core/backend/service/AbstractVFSDiagramService.java#L326) method as public static and reuseit... anyway I think it's better,as this `loadPath` method does, to delegate IO operations to the framework, if possible.

xiezhang7 replied:
@romartin , `loadPath` method is used to read a file, but here I don't use any file. I only need to convert a `String` to an `InputStream`, because the argument of the [unmarshall](https://github.com/kiegroup/kie-wb-common/blob/780c425886a399148819d023e5cb955df29dc707/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-project-backend/src/main/java/com/kie/workbench/common/stunner/cm/service/CaseManagementSwitchViewServiceImpl.java#L94) method is `InputStream`.

romartin replied:
ohh true, sorry! :+1:

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/util/CaseManagementUtils.java:
```diff
@@ -35,4 +60,111 @@
     }
+
+    @SuppressWarnings("unchecked")
+    public static int getGraphIndex(final Node parent,
```
the name of the method does not really describes what does it returns...

xiezhang7 replied:
Renamed to [getChildGraphIndex](https://github.com/kiegroup/kie-wb-common/blob/780c425886a399148819d023e5cb955df29dc707/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-client/src/main/java/org/kie/workbench/common/stunner/cm/client/command/util/CaseManagementCommandUtil.java#L52).

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/util/CaseManagementUtils.java:
```diff
@@ -35,4 +60,111 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    public static int getCanvasChildIndex(final Node<View<?>, Edge> parent,
```
I would prefer not to use the term "canvas" in classes inside the API modules...

xiezhang7 replied:
Moved the code to [client](https://github.com/kiegroup/kie-wb-common/blob/780c425886a399148819d023e5cb955df29dc707/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-client/src/main/java/org/kie/workbench/common/stunner/cm/client/command/util/CaseManagementCommandUtil.java#L68) package.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/util/CaseManagementUtils.java:
```diff
@@ -35,4 +60,111 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    public static int getCanvasNewChildIndex(final Node<View<?>, Edge> parent) {
```
I would prefer not to use the term "canvas" in classes inside the API modules...

xiezhang7 replied:
Moved the code to [client](https://github.com/kiegroup/kie-wb-common/blob/780c425886a399148819d023e5cb955df29dc707/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-client/src/main/java/org/kie/workbench/common/stunner/cm/client/command/util/CaseManagementCommandUtil.java#L126) package.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-project-client/src/main/java/org/kie/workbench/common/stunner/cm/project/client/editor/CaseManagementDiagramEditor.java:
```diff
@@ -104,6 +132,66 @@ public CaseManagementDiagramEditor(final AbstractProjectDiagramEditor.View view,
+                        final boolean toProcess = CaseManagementDiagramEditor.this.switchedToProcess.getAndSet(true);
+                        if (!toProcess) {
+                            if (!CaseManagementDiagramEditor.this.getProcessSessionPresenter().isPresent()) {
+                                CaseManagementDiagramEditor.this.setProcessEditorSessionPresenter(
```
I'm not sure if you're destroying properly the `SessionEditorPresenter` instance, once switching between them, by calling the `SessionEditorPresenter::destroy` method...

Because as far as I see the approach is having only a single `SessionEditorPresenter` instance in memory, and once switching views, it creates and sets a new one, is that correct? If so, can you please **ensure this code is not introducing any memory leak**? I'm talking due to my own experience too :) I introduced lots before hehe... so just please check the destroy method is being called  and all state is properly removed.

Just for brainstorming - Another approach would be to keep 2 instances of SessionEditorPresenter for this class state, one for the case view and another one for the process view. This has some benefits and downsides compared as your  actual implementation, but rather than this not sure if it will completely work due to the way Stunner manages "sessions", as just a single one active is expected at any time. So I don't think it worths now going this way, but was jsut throwing another idea if you want to give it a try.

xiezhang7 replied:
Added [code](https://github.com/kiegroup/kie-wb-common/blob/275abfeab02eae654f009ba2ec641fb39d3092e5/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-project-client/src/main/java/org/kie/workbench/common/stunner/cm/project/client/editor/CaseManagementDiagramEditor.java#L235) to call `SessionEditorPresenter::destroy`.

romartin commented:
Please @LuboTerifaj @hasys don't forget this PR, it's an important (and complex) new feature we're introducing for 7.4, so as soon as we can provide feedback to @xiezhang7 we'll obtain better results :)
Thanks in advance guys!

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Hi @romartin @LuboTerifaj , I added a second [commit](https://github.com/kiegroup/kie-wb-common/pull/2650/commits/59e5fa4eb2b9e48844cd0550a4b13cfec2367477) to support adding button to the tab bar. Thanks.

romartin commented:
great, thanks @xiezhang7 ! Probably also @manstis wants to take a look at [this commit](https://github.com/kiegroup/kie-wb-common/pull/2650/commits/59e5fa4eb2b9e48844cd0550a4b13cfec2367477) too, as it introduces a few changes on the "generic" editor side :+1:

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Hi @romartin ,

> Looks that the validate button does nothing?

There is a separate task for the validate button: [JBPM-8167](https://issues.jboss.org/browse/JBPM-8167)


> As now both BPMN and CM are being "merged" to share capabilities:
> 
> - I was expecting no beans (@Definition) declared in the CM API module
> - I was expecting no custom converters (marhsalling) declared in the CM backend module

Currently, I still cannot get rid of these things because there are some special properties for the case modeler element forms, e.g. Ad-Hoc autostart property. Please find the details in [JBPM-8035](https://issues.jboss.org/browse/JBPM-8035). 

Thanks.

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Hi @LuboTerifaj , I have fixed some of these issues and given explanation to the rest of the items. Thanks.

> - If there are two or more Stages created in case view, all except first one are placed on the same position in process view.

Will fix this when Process designer supports auto layout. 

> - When changing view a case/process is automatically saved. Auto-save feature has not been implemented yet and a user cannot enable or disable this behaviour, therefore it shouldn't be possible to do that neither with view change. After any change and switch to another view, the message says that case/process is saved when click on save button, but actually it is not. After close and open case/process, all changes are lost.

There is no auto-save. However I fixed an issue related saving diagram on Monday, I am not sure if it is related. If a diagram is not saved, after switching, it is still not saved. If a diagram is unchanged, it is still unchanged after switching. 
![Peek 2019-05-28 15-09](https://user-images.githubusercontent.com/42279951/58511937-d7962000-8169-11e9-89d5-794eef79ee00.gif)


> - If any node contain unsupported element in case view, the whole stage to which the node belongs (or stage itself) is not displayed.

I think if a node contains unsupported element, the node is not supported any more and the whole adhoc subprocess is not a stage any more. So it is not displayed.

> - If the ad-hoc sub-process (Stage) is added to the canvas in process view, after adding activities to the stage in case view, they are not placed in the upper left corner but in the lower left corner. If there is space for it, the activities should be added from the upper part of the ad-hoc sub-process to prevent wasting space.

This is fixed. Thanks.
![Peek 2019-05-28 16-37](https://user-images.githubusercontent.com/42279951/58511998-014f4700-816a-11e9-9854-331fb0834d0e.gif)


> - After changing to another version, it is not possible to switch between views. It is possible only after click on restore button.

This is fixed. Thanks.
![Peek 2019-05-28 16-38](https://user-images.githubusercontent.com/42279951/58512038-1b892500-816a-11e9-863c-1d1cd323cc48.gif)


> - Reordering Stages in case view created in process view without sequence flows causes broken sequence flows in process view. Just try to add ad-hoc sub-process to the process, switch to case view, change order of Stages and switch back.

I could not reproduce this issue.

> - It is not possible to open Diagram properties in case view.
>  -- "An unexpected error happened. Please try again in a few moments."
>  -- | Uncaught exception: CDI Event exception: CommandType=CDIEvent, BeanType=org.kie.workbench.common.stunner.core.client.canvas.event.selection.CanvasSelectionEvent, BeanReference=org.kie.workbench.common.stunner.core.client.canvas.event.selection.CanvasSelectionEvent@a5ba9, FromClient=1 sent to [unavailable] Caused by: (TypeError) : b is null
>  It also happens when any property is changed in process view, after switch to case view and click on a Stage.
>  If a new node is added in case view instead of clicking on the node, a message is displayed: "An error happened [com.google.gwt.core.client.JavaScriptException: (TypeError) : b is null]. Do you want to undo the last action?"
>  This can cause the whole business-central broken.

I have fixed the stage properties issue. I am not sure if all these issues are fixed because I could not reproduce all of them.
![Peek 2019-05-28 16-41](https://user-images.githubusercontent.com/42279951/58512620-8ab34900-816b-11e9-92b6-36fe784a9303.gif)
![Peek 2019-05-28 16-46](https://user-images.githubusercontent.com/42279951/58512636-91da5700-816b-11e9-9803-a3881b16fb61.gif)



> - Moving Stage to Lane breakes case view and after moving with nodes which were inside in this stage in case view, the outgoing sequence flow from the Stage has set target node to the "canvas" in process view. When moving with these node and switching between views, there are again unexpected error messages.

Lane is not yet supported for the switching, because I don't know how to handle the relationship between the stages in a lane and the stages not in a lane. I'd like to open a separate ticket to discuss the details.

> - When the target node of Stage is set to unsupported node, after switch between nodes, a source of the sequence flow is the "canvas".



> - Properties for nodes are different for case view in comparison with process view. Also ad-hoc property in case view is missing.

The difference is requested in task JBPM-8035. The adhoc property is requested to be hidden.

lazarotti commented:
@LuboTerifaj see that it is a tech preview component for 7.4 - so if it is not all issues fixed until Friday even though we should consider a merge and open Jiras to get the fixes further. Sure if the feature breaks totally the usage of Case modeler it should not be in, however if for particular use cases it does not work, so we can enhance and fix these scenarios further

romartin commented:
Hey @LuboTerifaj 
Yeah I'm also agree with @lazarotti , I don't see any blocker issues, considering that's tech preview. I'll proceed to test the generated WAR once the build completes, will test also the bpmn editor and ensure the changes to the marshallers are still working fine, in that case, I would agree on reporting the issues found as a new tickets and address those for next 7.5 release.

BTW About "auto-save" feature - I think that was not expected, IMO it also doesn't make sense, I see more intuitive the way it works now, so switching the view does NOT saves the process. What you think @krisv @lazarotti @ibek ? Anyway, as the other issues, that's not blocking, it can be reported and scheduled for 7.5 as well, np.

Thanks for all your work guys! :+1:

lazarotti commented:
IMO not saving should be the way to go. If something goes wrong after switching the views, at least the user can easily return to the previous state just closing the editor

xiezhang7 commented:
> - Regarding issues with sequence flows:
> 
>   -- After adding ad-hoc sub-process in process view, switching to case view and back, the sequence flows are not automatically added/reordered.
>   -- When you add more actvites to the stage, it can cause that sequence flows are broken in process view.
>   -- Steps to reproduce:
> 
> 1. Add ad-hoc sub-process in process view
> 2. Possibly add 6 or more activities to this sub-process
> 3. In case view change order of Stages
> 4. Switch views

Hi @LuboTerifaj , I still could not reproduce this issue. Please see the screencast below and let me know if anything is missing. Thanks.
![Peek 2019-05-29 13-15](https://user-images.githubusercontent.com/42279951/58577770-36fd3a00-8215-11e9-8c69-56861040a64a.gif)

xiezhang7 commented:
> Regarding the code, why did you remove some of the tests?

@LuboTerifaj , I removed some of the tests, because I moved or deleted the corresponding methods or classes.

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
@LuboTerifaj I have corrected the test names and test exceptions. Thanks.

xiezhang7 commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/factory/CaseManagementGraphFactoryImpl.java:
```diff
@@ -135,11 +149,39 @@ protected DefinitionManager getDefinitionManager() {
         commands.add(graphCommandFactory.addChildNode(diagramNode, stageNode));
+        commands.add(graphCommandFactory.addChildNode(diagramNode, endEventNode));
+        commands.add(graphCommandFactory.setSourceNode(startEventNode, startEventEdge,
+                                                       new MagnetConnection.Builder().atX(475d).atY(475d).auto(true).build()));
```
The numbers are for the initial locations of the Connections. Replaced with [constants](https://github.com/kiegroup/kie-wb-common/blob/78eaf5c03b5b49a00cad27f836fb10e42d70e1fe/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/factory/CaseManagementGraphFactoryImpl.java#L181).

xiezhang7 commented:
> The feature is not ready for usage because significant issue is still present there:
> 
> - If there are two or more Stages created in case view, all except first one are placed on the same position in process view.
> - Relating to the issue above, the behaviour is different than expected also in case, that you delete already existing Stage.
> 
> Steps to reproduce:
> 
> 1. Create second Stage in case view.
> 2. Delete first already existing one.
> 3. Switch the view.

@LuboTerifaj Although this is a significant issue, I don't think it stops users from using this feature as tech preview. Because users can manually move the stages to other locations in the process view. It is just missing the auto-layout capability.

xiezhang7 commented:
Jenkins execute full downstream build

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/property/task/AdHocSubprocessTaskExecutionSet.java:
```diff
@@ -57,7 +57,7 @@
     @SelectorDataProvider(
             type = SelectorDataProvider.ProviderType.CLIENT,
-            className = "org.kie.workbench.common.stunner.bpmn.client.dataproviders.AdHocOrderingProvider")
+            className = "org.kie.workbench.common.stunner.bpmn.client.dataproviders.ExecutionOrderProvider")
```
does this will affect actual BPMN editor in some way?

xiezhang7 replied:
@romartin I changed this because [AdHocSubprocessTaskExecutionSet](https://github.com/kiegroup/kie-wb-common/blob/e19070bbca3181721ac94f1517d9ba3fb73873f0/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/task/AdHocSubprocessTaskExecutionSet.java#L53) for BPMN was changed to use this value.

xiezhang7 commented:
Thanks @LuboTerifaj !

LuboTerifaj commented:
Reported issues: https://issues.jboss.org/browse/JBPM-8576?filter=-2&jql=issue%20in%20(JBPM-8560%2C%20JBPM-8561%2C%20JBPM-8563%2C%20JBPM-8565%2C%20JBPM-8568%2C%20JBPM-8569%2C%20JBPM-8571%2C%20JBPM-8576)
